GTFS profile & routing

takb opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [x] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

### Information about the changes
- Key functionality added:
GTFS routing support. 

Documentation needs to be updated before we roll out the next version. @koebi could maybe also help me check if the branch feature-gtfs  is fully superseded by this, please.

koebi committed:
add graphhopper gtfs-reader

takb committed:
enable GHGtfs class & profile declaration

takb committed:
added gtfs file for testing

koebi committed:
add API fields and methods for GTFS

koebi committed:
add conversion from route request to GraphHopper PTRequest

takb committed:
use ptRouter

koebi committed:
add pt request conversion to RoutingProfile

takb committed:
fix RouteResultBuilder

takb committed:
fix RouteResultBuilder errors

takb committed:
fix api test

takb committed:
fix pt instructions

takb committed:
include legs in RouteResult

takb committed:
pass legs to output WIP

takb committed:
finish leg output

takb committed:
test config

takb committed:
test elevation

takb committed:
Merge branch 'gh-4.0-upgrade' into gh-4.0-gtfs

takb committed:
use GH fork release, API test

takb committed:
gh-gtfs-4.0.1 stop storing pt edge names

takb committed:
gh-gtfs-4.0.2 use GH 5 GTFS impl

takb committed:
gh-gtfs-4.0.3 move transfer parsing in PtRouterImpl.Factory

takb committed:
gh-gtfs-4.0.4 fix pt leg output & instructions, geometry fix WIP

takb committed:
gh-gtfs-4.0.5 implemented GTFS shape matching and output

takb committed:
minor typo in language resource

takb committed:
gh-gtfs-4.0.6 fix geojson output, instructions indexing, handling of empty shapes

takb committed:
gh-gtfs-4.0.7 fix pt distance caluculation

takb committed:
detect cases where entry and exit PT stations cannot be reached within street time

takb committed:
add error codes and route_type/route_desc output

takb committed:
gh-gtfs-4.0.9 config parameter for max_visited_nodes, improved error messages

koebi committed:
set defaults for walking_time, schedule and ignore_transfers

takb committed:
Merge branch 'master' into gh-4.0-gtfs

takb committed:
feat(GTFS): Add GTFS profile support

Enables GTFS profile for routing

takb committed:
Merge branch 'master' into gh-4.0-gtfs

takb committed:
CI test memory option

takb committed:
doc(pt): hide parameters in swagger

MichaelsJP commented on CHANGELOG.md:
```diff
@@ -34,6 +34,8 @@ RELEASING:
 
 ## [Unreleased]
+### Added
+- GTFS support
```
```suggestion
- GTFS support for routing  ([#1316](https://github.com/GIScience/openrouteservice/pull/1316))
```

MichaelsJP commented on .github/workflows/run_maven_tests.yml:
```diff
@@ -35,7 +35,7 @@ jobs:
       run: cp ./openrouteservice-api-tests/conf/ors-config-test.json ./openrouteservice/src/main/resources/ors-config.json
     - name: Test, build, run API tests
-      run: mvn -B -f openrouteservice/pom.xml verify jacoco:report
+      run: export MAVEN_OPTS="-Xmx3000m -Xms2500m" && mvn -B -f openrouteservice/pom.xml verify jacoco:report
```
We should probably use a smaller test set?

takb replied:
done

MichaelsJP commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java:
```diff
@@ -3651,6 +3653,177 @@ public void expectMaxpeedHgvForward() {
+            .body("features[0].properties.legs[1].containsKey('route_desc')", is(true))
+            .body("features[0].properties.legs[1].containsKey('geometry')", is(true))
+            .statusCode(200).extract().response();
+    }
```
```suggestion
    @Test
    public void testPTJSON() {
        JSONArray coordinates =  new JSONArray();
        JSONArray coord1 = new JSONArray();
        coord1.put(8.6729581);
        coord1.put(49.4468535);
        coordinates.put(coord1);
        JSONArray coord2 = new JSONArray();
        coord2.put(8.7067204);
        coord2.put(49.3786147);
        coordinates.put(coord2);
        JSONObject body = new JSONObject();
        body.put("coordinates", coordinates);
        body.put("instructions", true);
        body.put("elevation", true);
        body.put("departure", "2022-07-04T13:02:26Z");
        body.put("walking_time", "PT30M");
        Response res = given()
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .pathParam("profile", getParameter("ptProfile"))
            .body(body.toString())
            .when()
            .post(getEndPointPath() + "/{profile}")
            .then().log().ifValidationFails()
            .assertThat()
            .body("any { it.key == 'routes' }", is(true))
            .body("routes.size()", is(3))
            .body("routes[0].summary.transfers", is(2))
            .body("routes[0].legs.size()", is(5))
            .body("routes[0].legs[0].containsKey('arrival')", is(true))
            .body("routes[0].legs[0].containsKey('departure')", is(true))
            .body("routes[0].legs[0].containsKey('instructions')", is(true))
            .body("routes[0].legs[0].containsKey('geometry')", is(true))
            .body("routes[0].legs[1].containsKey('feed_id')", is(true))
            .body("routes[0].legs[1].containsKey('trip_id')", is(true))
            .body("routes[0].legs[1].containsKey('route_id')", is(true))
            .body("routes[0].legs[1].containsKey('route_type')", is(true))
            .body("routes[0].legs[1].containsKey('route_desc')", is(true))
            .body("routes[0].legs[1].containsKey('geometry')", is(true))
            .statusCode(200).extract().response();
    }

    @Test
    public void testPTGeoJSON() {
        JSONArray coordinates =  new JSONArray();
        JSONArray coord1 = new JSONArray();
        coord1.put(8.6729581);
        coord1.put(49.4468535);
        coordinates.put(coord1);
        JSONArray coord2 = new JSONArray();
        coord2.put(8.7067204);
        coord2.put(49.3786147);
        coordinates.put(coord2);
        JSONObject body = new JSONObject();
        body.put("coordinates", coordinates);
        body.put("instructions", true);
        body.put("elevation", true);
        body.put("departure", "2022-07-04T13:02:26Z");
        body.put("walking_time", "PT30M");
        Response res = given()
            .header("Accept", "application/geo+json")
            .header("Content-Type", "application/json")
            .pathParam("profile", getParameter("ptProfile"))
            .body(body.toString())
            .when()
            .post(getEndPointPath() + "/{profile}/geojson")
            .then().log().ifValidationFails()
            .assertThat()
            .body("any { it.key == 'features' }", is(true))
            .body("features.size()", is(3))
            .body("features[0].properties.transfers", is(2))
            .body("features[0].properties.legs.size()", is(5))
            .body("features[0].properties.legs[0].containsKey('arrival')", is(true))
            .body("features[0].properties.legs[0].containsKey('departure')", is(true))
            .body("features[0].properties.legs[0].containsKey('instructions')", is(true))
            .body("features[0].properties.legs[0].containsKey('geometry')", is(true))
            .body("features[0].properties.legs[1].containsKey('feed_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('trip_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_id')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_type')", is(true))
            .body("features[0].properties.legs[1].containsKey('route_desc')", is(true))
            .body("features[0].properties.legs[1].containsKey('geometry')", is(true))
            .statusCode(200).extract().response();
    }
```

MichaelsJP commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java:
```diff
@@ -3651,6 +3653,177 @@ public void expectMaxpeedHgvForward() {
+    }
+
+    @Test
+    public void testPTFail() {
```
Either split in multiple tests or parametrize the tests.

takb replied:
Parametrized.

takb committed:
Update CHANGELOG.md

Co-authored-by: Julian Psotta <22392104+MichaelsJP@users.noreply.github.com>

takb committed:
Update openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java

Co-authored-by: Julian Psotta <22392104+MichaelsJP@users.noreply.github.com>

takb committed:
ci(GTFS): use smaller GTFS dataset for API testing

replaced example GTFS dataset with a subset to reduce memory footprint when run on ors-config-test.json setup.

As suggested here https://github.com/GIScience/openrouteservice/pull/1316#discussion_r1131290450

takb committed:
test(GTFS): parametrize testPTFail

Required updating junit and adding junit-jupiter-params.

TheGreatRefrigerator commented:
some issues with the gtfs documentation:
- [x] `pt` profile should prob. be `public-transport` as it doesn't fit with the rest.
- [ ] `pt` profile should only be used with directions (if not implemented for isos and matrix). Not sure if we can restrict profiles for specific endpoints yet
- [x] default and example values for `ignore_transfers`, `schedule`, `schedule_duration`, `schedule_rows` and `walking_time` should be meaningful. Default values should be preferred over example values so the info "what is set when the param is not passed" is there.
- [x] the description(`value` in `ApiModelProperty`) for the above parameters doesn't really make clear what happens if you set them. Probably a bit more explanation on what they do.
- [x] the description sometimes contains `public transit` code always names things `public_transport` (except for `PUBLIC_TRANSIT` in APIEnums.java) -> use `public_transport` consistently and get rid of public transit, also in comments

TheGreatRefrigerator committed:
fix(swagger): generated swagger splits for request & response

the automatic swagger.json generation was creating some duplicate entries
since the 3.0.0 update of springfox-swagger (e.g. Directions ServiceReq &
Directions ServiceRes instead of Directions Service).
In addition this created some circular dependencies which break the
API Playground as it needs all references resolved.

The issue is described here https://github.com/springfox/springfox/issues/3487#issuecomment-1134579077
and this commit contains the solution from there to prevent the split.

takb committed:
doc(GTFS): re-enable API fields for swagger

enable GTFS-related RouteRequest fields for swagger representation

koebi commented:
- [x] add config documentation for gtfs profile
- [ ] add info that transfers will be calculated by a set foot profile

koebi committed:
fix: consistently name public transport profile

To fit with all other profiles, pt is renamed to public transport.
This is also applied to any places mentioning transit.

koebi committed:
docs(API): add default values for public transport params

koebi committed:
docs(BE): add public transport config values

takb committed:
Merge branch 'master' into gh-4.0-gtfs

MichaelsJP commented on openrouteservice/src/main/java/org/heigit/ors/api/responses/routing/json/JSONLeg.java:
```diff
@@ -0,0 +1,138 @@
+@ApiModel(value="JSONLeg", description = "Leg of a route")
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public class JSONLeg {
+    @ApiModelProperty(value = "The type of the leg, possible values are currently 'walk' and 'pt'.", example = "pt")
```
Shouldn't we better rename all `pt` instances to `public-transit` ?

takb replied:
![image](https://user-images.githubusercontent.com/31958783/224936627-24b61ec9-a441-4192-bdc5-d181beb2a17b.png)
@koebi renamed `pt`  to `public-transport` and it seems okay to me...

MichaelsJP commented on openrouteservice-api-tests/src/test/java/org/heigit/ors/v2/services/routing/ResultTest.java:
```diff
@@ -110,6 +140,7 @@ public ResultTest() {
         addParameter("bikeProfile", "cycling-regular");
         addParameter("carProfile", "driving-car");
         addParameter("footProfile", "foot-walking");
+        addParameter("ptProfile", "pt");
```
```suggestion
        addParameter("ptProfile", "public-transit");
```

MichaelsJP commented on openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfileType.java:
```diff
@@ -192,6 +194,9 @@ public static int getFromString(String profileType) {
             case "wheelchair":
                 return WHEELCHAIR;
 
+            case "pt":
```
```suggestion
            case "public-transit":
```

takb committed:
fix(GTFS): profile name in enum

Fix profile name in RoutingProfileType, fix API test with hardcoded profile name

takb committed:
Merge branch 'master' into gh-4.0-gtfs

takb commented:
@koebi says he's done with docs and checks, I think this is ready @MichaelsJP
